### ISSUE TITLE
`f32::from(<untyped float>)` inference with `f16` and `f128`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust#123831]
Adding a `From<f16> for f32` impl breaks some inference that currently works. From https://github.com/rust-lang/rust/issues/123824:

```rust
fn main() {
    let x = f32::from(3.14);
}
``` 

The new float types are a long way from being stable, but it may be good to decide how to handle things if we would like this implementation one day. Per @fmease at https://github.com/rust-lang/rust/issues/123824#issuecomment-2050628184, there are three options:

1. Accept this regression since technically, no stable API is broken
2. Say that we do not want `From<f16> for f32`
3. Make this implementation available starting with the new edition

I'll mark this with the edition label so it can at least come into consideration, but this is the lowest priority thing on any list.

Known failure points from adding this directly:
- Fuchsia: https://cs.opensource.google/fuchsia/fuchsia/+/main:src/recovery/lib/recovery-ui/src/keyboard.rs;l=111;drc=e1ed8a47e77b0ea1d6f83b93830c4e975a8a6089 (from https://github.com/rust-lang/rust/issues/123824#issuecomment-2050836467)
-  `taffy`: https://github.com/DioxusLabs/taffy/blob/f1e4edab96ee13f7d1fbc2536ac31633b0e620d6/src/style_helpers.rs#L58 (from https://github.com/rust-lang/rust/issues/123824#issuecomment-2052506991)
- `contour-rs`: https://github.com/mthh/contour-rs/blob/08c1d83a1461aa19543ed402d23908eb55234fd1/src/lib.rs#L488-L491 (from https://github.com/mthh/contour-rs/pull/13#issuecomment-2051652708)
- Bevy: https://github.com/rust-lang/rust/issues/123831#issuecomment-2052981580
- egui: https://github.com/emilk/egui/issues/4352

@rustbot label +T-lang +T-libs +A-edition-2024 +F-f16_and_f128

**Comment by compiler-errors** [comment]
There's no way we can provide an impl conditional on edition, since those are always global and can't really be conditional based on the way that the trait solver works. The way that this worked for `into_iter()` and arrays is that it only affected dot-method syntax, so we just skipped over the candidate for `[T; N]` when we had `[].into_iter()` -- I expect this to have to be implemented differently for `From::from` and/or `f32::from`.

**Comment by tgross35** [comment]
Too bad, but that makes enough sense. I guess it would be interesting to see crater results, though the fact that this was discovered within 24 hours of merge makes me think there could be quite a lot of codebases relying on this somewhere...

@rustbot label -A-edition-2024

**Comment by tgross35** [comment]
Crater run at https://github.com/rust-lang/rust/pull/122470#issuecomment-2050824749

**Comment by workingjubilee** [comment]
Some languages give literals their own special type.

**Comment by nicoburns** [comment]
Some kind of improvement to type inference such that it can handle this case would same ideal. Perhaps even a lang feature to allow user code to guide the inference process.

**Comment by BD103** [comment]
Can the link to the Taffy regression be updated to https://github.com/DioxusLabs/taffy/blob/f1e4edab96ee13f7d1fbc2536ac31633b0e620d6/src/style_helpers.rs#L58, before it was fixed in https://github.com/DioxusLabs/taffy/commit/9f27870656a06f125f63b46e37c2038fcca1818e?

**Comment by BD103** [comment]
I'm also getting the same warnings for [this line]https://github.com/bevyengine/bevy/blob/5caf085dacf74bf553a0428a5eb7f4574a9bb99c/examples/2d/sprite_slice.rs#L9), [this line](https://github.com/bevyengine/bevy/blob/5caf085dacf74bf553a0428a5eb7f4574a9bb99c/examples/ui/grid.rs#L8), [this line](https://github.com/bevyengine/bevy/blob/5caf085dacf74bf553a0428a5eb7f4574a9bb99c/examples/ui/viewport_debug.rs#L28), [this line](https://github.com/bevyengine/bevy/blob/5caf085dacf74bf553a0428a5eb7f4574a9bb99c/examples/window/window_settings.rs#L18), and [these lines](https://github.com/bevyengine/bevy/blob/5caf085dacf74bf553a0428a5eb7f4574a9bb99c/crates/bevy_ui/src/layout/convert.rs#L573-L584).

Most of them are related to casting a tuple `({float}, {float})` / array `[{float}; 2]` to our custom [`WindowResolution`](https://docs.rs/bevy/0.13.2/bevy/window/struct.WindowResolution.html) struct. The last one is different, though, because it runs into Taffy's `fn foo<T: Into<f32>>(x: T)` functions.

I hope this helps :)

**Comment by tgross35** [comment]
Crater finished https://crater-reports.s3.amazonaws.com/pr-122470/index.html

**Comment by compiler-errors** [comment]
Hm -- looks like a lot of these are due to `Into<f32>` bounds (e.g. in arguments) and not due to method dispatch.

As I stated above, impls *need* to apply globally, so it's gonna be a bit harder to fix this (if possible at all, without hacking out something that I'd be afraid of having to support in perpetuity).

**Comment by tgross35** [comment]
Yeah, I would rather not have this impl than have something specific and hacky.

How feasible would it be for untyped literals to select their type based on the bound? As in, if there is a requirement for `impl Into<T>` and an unsuffixed literal could represent `T`, then do that rather than falling back to the default `f64`/`i32`. Which I think would resolve most of the failures here, in addition to making something like this work:

```rust
fn foo(a: impl Into<u64>) {}

fn main() {
    foo(100);
}
```
